homeassistant/components/nextcloud/update.py

```diff
@@ -0,0 +1,53 @@
+"""Summary data from Nextcoud."""
+from __future__ import annotations
+
+from homeassistant.components.update import UpdateEntity, UpdateEntityDescription
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DOMAIN
+from .coordinator import NextcloudDataUpdateCoordinator
+from .entity import NextcloudEntity
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the Nextcloud update entity."""
+    coordinator: NextcloudDataUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
+    async_add_entities(
+        [
+            NextcloudUpdateSensor(
+                coordinator, entry, UpdateEntityDescription(key="update")
+            )
+        ]
+    )
+
+
+class NextcloudUpdateSensor(NextcloudEntity, UpdateEntity):
+    """Represents a Nextcloud update entity."""
+
+    @property
+    def installed_version(self) -> str | None:
+        """Version installed and in use."""
+        return self.coordinator.data.get("system_version")
+
+    @property
+    def latest_version(self) -> str | None:
+        """Latest version available for install."""
+        return self.coordinator.data.get(
+            "system_update_available_version",
+            self.installed_version
+            # needed for Nextcloud serverinfo app pre 1.18.0
+            if any(x.startswith("system_update") for x in self.coordinator.data)
+            else None,
+        )
```

Comment:
```suggestion
        return self.coordinator.data.get(
            "system_update_available_version", self.installed_version
        )
```

this special handling seems to be unnecessary, since `self.installed_version` will already return a valid string or None.
Further the current [serverinfo](https://github.com/nextcloud/serverinfo/tags) version is 28.0.1, so 1.18.0 seems to a very long time ago 🤔 

Nevertheless, my nextcloud is on 27.1.4 and the administration overview shows 27.1.5 available
![image](https://github.com/home-assistant/core/assets/35783820/772af3fc-df54-4d9f-bafd-f99628ad17c7)

but no information about an avaialble update is in the API output of the serviceinfo app
![image](https://github.com/home-assistant/core/assets/35783820/eea68950-26c7-4e06-8e04-5b983a0d1155)

what version of nextcloud do you use and how did you test this working?

Reply:
> ```suggestion
>         return self.coordinator.data.get(
>             "system_update_available_version", self.installed_version
>         )
> ```
> 
> this special handling seems to be unnecessary, since `self.installed_version` will already return a valid string or None.

But this is the currently installed version, which is used as you describe in the method above. Here we have to gather the last available one.

> Further the current [serverinfo](https://github.com/nextcloud/serverinfo/tags) version is 28.0.1, so 1.18.0 seems to a very long time ago 🤔 

Thia version does match the version of Nextcloud server, but the add-on uses another internally: https://github.com/nextcloud/serverinfo/blob/v27.1.4/appinfo/info.xml#L8

> 
> Nevertheless, my nextcloud is on 27.1.4 and the administration overview shows 27.1.5 available
> ![image](https://github.com/home-assistant/core/assets/35783820/772af3fc-df54-4d9f-bafd-f99628ad17c7)
> 
> but no information about an avaialble update is in the API output of the serviceinfo app
> ![image](https://github.com/home-assistant/core/assets/35783820/eea68950-26c7-4e06-8e04-5b983a0d1155)
> 
> what version of nextcloud do you use and how did you test this working?

That's the whole point. With Nextcloud 28 (add-on version 1.18) this information was added. The structure could be seen [here](https://github.com/nextcloud/serverinfo/blob/v28.0.0/lib/SystemStatistics.php#L81) and the method below.

Reply:
int this case I think we should only create the update entity in case the `system_update` data block available.
Without out it, the whole update entity is useless and the already existing `system_version` sensor should be enough.

Reply:
But the first time we know about the version is after the first coordinator update, which should happen after the `async_setup_entry` call. So, how to handle this right?

Reply:
the initial data update is done during the setup of the whole integration
https://github.com/home-assistant/core/blob/83b09341387f47653745a53edb01ebb991ea652a/homeassistant/components/nextcloud/__init__.py#L71
only than the platforms are going to setup, so during platform setup we already have "fresh" data in the coordinator.